### PR TITLE
Fix afni contrast bool

### LINKED ISSUE #265: AFNI estimator fails if contrasts have nested names
If one contrast has a name that is a substring of another contrast the AFNI estimator will currently fail to extract all of the correct maps due to a bug in how subbricks in the output of remlfit are being extracted. The failure occurs in `collate_l1_outputs` with `ValueError: List lengths must be consistent across fields`. I think I've got a fix for this and will submit the PR soon.

### PR DESCRIPTION
**Pull request by Shotgunosine**
closes #265

### CODE REVIEW

File: fitlins/interfaces/afni.py [hunk 1]
```diff
@@ -265,14 +265,24 @@ def save_remlfit_results(self, maps, contrasts, runtime):
         vol_labels = parse_afni_ext(maps["effect_size"])["BRICK_LABS"].split("~")
 
         effect_bool = np.array([x.endswith("Coef") for x in vol_labels])
+        clean_vol_labels = []
+        for ii, x in enumerate(vol_labels):
+            if (ii in [0, 1]):
+                clean_vol_labels.append(x)
+            elif x.endswith('_Coef') or x.endswith('_Tstat'):
+                clean_vol_labels.append('#'.join(x.split('#')[:-1]))
```
Comment (effigies) [inline fitlins/interfaces/afni.py:273]:
```suggestion
                clean_vol_labels.append(x.rsplit('#', 1)[0])
```

File: fitlins/interfaces/afni.py [hunk 2]
```diff
@@ -265,14 +265,24 @@ def save_remlfit_results(self, maps, contrasts, runtime):
...
+            elif x.endswith('_R^2') or x.endswith('_Fstat'):
+                clean_vol_labels.append('_'.join(x.split('_')[:-1]))
```
Comment (effigies) [inline fitlins/interfaces/afni.py:275]:
```suggestion
                clean_vol_labels.append(x.rsplit('_')[0])
```

File: fitlins/interfaces/afni.py [hunk 3]
```diff
@@ -265,14 +265,24 @@ def save_remlfit_results(self, maps, contrasts, runtime):
...
+                clean_vol_labels.append(x)
         for (name, weights, contrast_type) in contrasts:
             contrast_metadata.append(
                 {"contrast": name, "stat": contrast_type, **out_ents}
             )
 
             # Get boolean to index appropriate values
             stat_bool = stat_types == contrast_type.upper()
-            contrast_bool = np.array([((name in x) and (('stim_' + name ) not in x)) for x in vol_labels])
+            contrast_bool = np.array([name == x for x in clean_vol_labels])
```
Comment (effigies) [inline fitlins/interfaces/afni.py:285]:
```suggestion
            contrast_bool = np.array(clean_vol_labels) == name
```

File: fitlins/interfaces/afni.py [hunk 4]
```diff
@@ -265,14 +265,24 @@ def save_remlfit_results(self, maps, contrasts, runtime):
         vol_labels = parse_afni_ext(maps["effect_size"])["BRICK_LABS"].split("~")
 
         effect_bool = np.array([x.endswith("Coef") for x in vol_labels])
+        clean_vol_labels = []
+        for ii, x in enumerate(vol_labels):
+            if (ii in [0, 1]):
+                clean_vol_labels.append(x)
+            elif x.endswith('_Coef') or x.endswith('_Tstat'):
```
Comment (effigies) [inline fitlins/interfaces/afni.py:272]:
```suggestion
        clean_vol_labels = vol_labels[:2]
        for x in vol_labels[2:]:
            if x.endswith('_Coef') or x.endswith('_Tstat'):
```

**Review by effigies — COMMENTED** [review]
No obvious problems with the logic, but here are some simplifications.

File: fitlins/interfaces/afni.py (hunk 2 above)
Comment (Shotgunosine) [reply fitlins/interfaces/afni.py:275]: I think this should probably be x.rsplit('_', 1)[0] as well, right?

**Comment by effigies** [comment]
Tests passing. Is this branch working for you? Do you have a test model snippet that can exercise this?

**Comment by Shotgunosine** [comment]
Double checking the changes now. 

here's a slightly tweaked ds003_model001 that would provoke the bug if it was still present:

```
{
  "Name": "ds003_model001",
  "Description": "",
  "Input": {
    "task": "rhymejudgment"
  },
  "Steps": [
    {
      "Level": "run",
      "Transformations": [
        {
          "Name": "Factor",
          "Input": [
            "trial_type"
          ]
        },
        {
          "Name": "Convolve",
          "Input": ["trial_type.word", "trial_type.pseudoword"],
          "Model": "spm"
        }
      ],
      "Model": {
        "X": [
          "trial_type.word",
          "trial_type.pseudoword",
          "framewise_displacement",
          "trans_x", "trans_y", "trans_z", "rot_x", "rot_y", "rot_z",
          "a_comp_cor_00", "a_comp_cor_01", "a_comp_cor_02",
          "a_comp_cor_03", "a_comp_cor_04", "a_comp_cor_05"
        ]
      },
      "DummyContrasts": {
        "Conditions": ["trial_type.word", "trial_type.pseudoword"],
        "Type": "t"
        },
      "Contrasts": [
        {
          "Name": "word_gt_pseudo",
          "ConditionList": [
            "trial_type.word",
            "trial_type.pseudoword"
          ],
          "Weights": [1, -1],
          "Type": "t"
        },
        {
          "Name": "word_gt_pseudo_rev",
          "ConditionList": [
            "trial_type.word",
            "trial_type.pseudoword"
          ],
          "Weights": [-1, 1],
          "Type": "t"
        },
        {
          "Name": "task_vs_baseline",
          "ConditionList": [
            "trial_type.word",
            "trial_type.pseudoword"
          ],
          "Weights": [0.5, 0.5],
          "Type": "t"
        }
      ]
    },
    {
      "Level": "dataset",
      "DummyContrasts": {
        "Conditions": ["word_gt_pseudo", "task_vs_baseline"],
        "Type": "t"
        },
      "Contrasts":[
        {
          "Name": "any_words",
          "ConditionList": [
            "trial_type.word",
            "trial_type.pseudoword"
          ],
          "Weights": [
            [1, 0],
            [0, 1]
          ],
          "Type": "F"
        }
      ]
    }
  ]
}
```

As for that line that's not getting tested, I don't think that case should normally come up, but I didn't want the counts to get out of wack if it did come up for some reason.

**Comment by Shotgunosine** [comment]
It's working locally for me, I'd be ok if you wanted to go ahead and merge it.

**Comment by effigies** [comment]
Sounds good. Looks like we're covering the cases. Would be good to set up more rigorous tests when we have time.

**Pull request merged**